Solutions/Exercises/ex5.lean

```diff
@@ -0,0 +1,248 @@
+
+
+inductive myEven : Nat → Prop
+| zero : myEven 0
+| two_plus_n : ∀ n, myEven n → myEven (n + 2)
+
+inductive myOdd : Nat → Prop
+| one : myOdd 1
+| two_plus_n : ∀ n, myOdd n → myOdd (n + 2)
+
+/-! ### basic properties of myEven and myOdd -/
+
+
+open myEven myOdd
+
+-- reordered results
+theorem myOdd_succ_myEven : ∀ n, myOdd n → myEven (n + 1) := λ n hn => by
+induction hn with
+| one => exact two_plus_n 0 zero
+| two_plus_n n _ hhn => exact two_plus_n _ hhn
+
+theorem myEven_succ_myOdd : ∀ n, myEven n → myOdd (n + 1) := λ n hn => by
+induction hn with
+| zero => exact myOdd.one
+| two_plus_n n _ hhn => exact two_plus_n _ hhn
+
+theorem myEven_or_myOdd : ∀ n, myEven n ∨ myOdd n := λ n => by induction n with
+| zero => exact Or.inl zero
+| succ _ hd =>
+  rcases hd with (hd|hd)
+  { exact Or.inr (myEven_succ_myOdd _ hd) }
+  { exact Or.inl (myOdd_succ_myEven _ hd) }
+
+/-! ### addition -/
+
+theorem myEven_add_myOdd : ∀ m n, myEven m → myOdd n → myOdd (m + n) := λ m n hm hn => by
+induction hn with
+| one => exact myEven_succ_myOdd _ hm
+| two_plus_n _ _ hm => rw [← add_assoc]; exact two_plus_n _ hm
+
+theorem myOdd_add_myEven : ∀ m n, myOdd m → myEven n → myOdd (m + n) := λ m n hm hn => by
+rw [add_comm]; exact myEven_add_myOdd _ _ hn hm
+
+theorem myOdd_add_myOdd : ∀ m n, myOdd m → myOdd n → myEven (m + n) := λ m n hm hn => by
+induction hn with
+| one => exact myOdd_succ_myEven _ hm
+| _ _ _ hm => rw [← add_assoc]; exact two_plus_n _ hm
+
+theorem myEven_add_myEven : ∀ m n, myEven m → myEven n → myEven (m + n) := λ m n hm hn => by
+induction hn with
+| zero => exact hm
+| _ _ _ hn => rw [← add_assoc]; exact two_plus_n _ hn
+
+/-! ### multiplication -/
+
+lemma myEven_two_mul {n : ℕ} :
+  myEven (2 * n) := by
+induction n with | zero => exact zero | _ _ hd => rw [Nat.mul_succ]; exact two_plus_n _ hd
+
+theorem myEven_mul : ∀ m n, myEven m → myEven (m * n) := λ m n hm => by
+induction hm with
+| zero => rw [zero_mul]; exact zero
+| _ _ _ hn => rw [add_mul]; exact myEven_add_myEven _ _ hn myEven_two_mul
+
+theorem mul_myEven : ∀ m n, myEven n → myEven (m * n) := λ m n hn => by
+rw [mul_comm]; exact myEven_mul _ _ hn
+
+theorem myOdd_mul_myOdd : ∀ m n, myOdd m → myOdd n → myOdd (m * n) := λ m n hm hn => by
+induction hn with
+| one => rw [mul_one]; exact hm
+| _ _ _ hn => rw [mul_add, mul_comm _ 2]; exact myOdd_add_myEven _ _ hn myEven_two_mul
+
+/-! ### power -/
+
+-- forgot to say that `n : ℕˣ`
+theorem myEven_pow : ∀ m (n : ℕˣ), myEven m → (myEven (m ^ (n : ℕ))) := λ m n hm => by
+-- changing `n : ℕˣ` to `n' : ℕ` instead of having coercions everywhere
+let n' : ℕ := n; have : n' = n := rfl; simp only [← this, Ne.def] at *; clear this
+have hnn : 1 ≤ n' := Nat.one_le_iff_ne_zero.mpr (n.ne_zero)
+-- `revert` so that `n'` can be replaced in `hnn` as well
+revert hnn
+induction n' with
+| zero => intro; contradiction
+| succ _ _ => rw [pow_succ']; exact λ _ => mul_myEven _ _ hm
```

Comment:
Not sure how this happened, but this line is now giving me an error. Maybe it's because I updated mathlib? But that shouldn't give a type error here.

Reply:
My errors at the moment are with all imports. It's weird because the files in a different repository are fine. Something is up with the repo

Reply:
ya u right